[AVR] Insert JMP for long branches

exscape opened the pull request:
Previously, on long branches (relative jumps of >4 kB), an assertion
failure was hit, as `AVRInstrInfo::insertIndirectBranch` was not
implemented. Despite its name, it is called by the branch relaxator
for *all* unconditional jumps.

I have used this for a few days without any problems, and the tests pass.  
Please check the changes thoroughly, in case I've missed something important.

exscape committed:
[AVR] Insert JMP for long branches

Previously, on long branches (relative jumps of >4 kB), an assertion
failure was hit, as AVRInstrInfo::insertIndirectBranch was not
implemented. Despite its name, it is called by the branch relaxator
for *all* unconditional jumps.

exscape commented on lib/Target/AVR/AVRInstrInfo.cpp:
```diff
@@ -537,8 +537,7 @@ bool AVRInstrInfo::isBranchOffsetInRange(unsigned BranchOp,
   case AVR::CALLk:
-    assert(BrOffset >= 0 && "offset must be absolute address");
-    return isUIntN(16, BrOffset);
+    return true;
```
I think this should work, since these instructions should always be able to reach any address, from any address.  
Regardless, the previous `assert` appears incorrect: `BrOffset` is negative for all jumps/calls where the destination address is lower to the address of the `JMP`/`CALL` instruction, which should certainly be valid.

dylanmckay replied:
I don't think that is quite true

AVR supports addressing of >16-bit pointers through the use of extended IO instructions. In these cases, bits can be written to a special IO register which will then be taken to be the high bits of the address for an IO operation.

Also, LLVM will likely generate >16kb branches if the program is big enough. For example, if we had 16kb of NOPs and a call to a symbol located afterwards, LLVM might possibly make it all the way to this line before erroring.

Although I do have a feeling that we probably have an assert for this somewhere earlier in the pipeline, I think this assertion should stay here

dylanmckay replied:
[Here's](https://imagecraft.com/help/iccavr/wwhelp/wwhimpl/common/html/wwhelp.htm?context=ICCAVRHelp&file=6-ProgrammingAVR12.html) some documentation on the extended IO feature

exscape replied:
I'll read up on that and test bigger branches, but note that I didn't remove those lines without reason. If the logic needs to remain, they need fixing, and I'm not sure how.

With the call to `isUIntN` left in, the code enters an infinite loop that uses up all RAM by creating hundreds of thousands of basic blocks, as `isBranchOffsetInRange` returns false when inserting a backwards jump (`BrOffset` negative), which makes it insert another.  
With the `assert` left in, we get an assertion failure any time we jump to a lower PC.

exscape replied:
Are you sure that's relevant? It looks to be about data (including LPM) access only, not instructions.  
JMP should, according to the AVR Instruction Set Manual, be able to reach any location on any AVR (up 4M words of memory), as the address is stored as 22 bits.

As for CALL, the manual shows it as having two forms, for 16-bit vs 22-bit PC, but there's only one opcode, with 22 bits.  
Each versions handles the entire program memory for the devices that has that instruction form,128 kB for devices with 128 kB and less, up to 8 MB for devices with more than 128 kB).

I successfully tested compiling and linking a .ll file with a ~30 kB jump with no issues, by the way. 😃

dylanmckay replied:
It seems like I'm crossing my wires!

I believe you are correct, nice investigating

exscape commented on test/CodeGen/AVR/branch-relaxation-long.ll:
```diff
@@ -0,0 +1,4133 @@
+  call void asm sideeffect "nop", ""()
+  call void asm sideeffect "nop", ""()
+  call void asm sideeffect "nop", ""()
+  call void asm sideeffe
```
The second test (of two) begins here.

dylanmckay commented on lib/Target/AVR/AVRInstrInfo.h:
```diff
@@ -107,6 +107,12 @@ class AVRInstrInfo : public AVRGenInstrInfo {
+  unsigned insertIndirectBranch(MachineBasicBlock &MBB,
+                                MachineBasicBlock &NewDestBB,
+                                const DebugLoc &DL,
+                                int64_t BrOffset = 0,
```
I would avoid specifying defaults here so there's no risk we get out of sync with upstream LLVM.

I doubt that there will any by callers of `AVRInstrInfo::insertIndirectBranch` directly - the only caller will likely be outside of `lib/Target/AVR` and thus will automatically see the default values because it will be working through a `TargetInstrInfo*` object.

dylanmckay commented on test/CodeGen/AVR/branch-relaxation-long.ll:
```diff
@@ -0,0 +1,4133 @@
+; CHECK: jmp     LBB0_2
+
+define i8 @relax_to_jmp(i1 %a) {
+entry-block:
```
Add `CHECK-LABEL: LBB0_1` above this label.

If somehow the basic block default name numbering changed, this will catch that

dylanmckay commented on test/CodeGen/AVR/branch-relaxation-long.ll:
```diff
@@ -0,0 +1,4133 @@
+define i8 @relax_to_jmp(i1 %a) {
+entry-block:
+  br i1 %a, label %hello, label %finished
+hello:
```
Add `CHECK-LABEL: LBB0_2` above this label

dylanmckay reviewed (changes requested):
Looks really nice! just a few nitpicks left, along with adding the assertion back in

exscape commented:
Hmm, I've fixed the things you commented on, but I noticed that `CodeGen/AVR/high-pressure-on-ptrregs.ll` fails (`llc: .../include/llvm/CodeGen/LiveIntervalAnalysis.h:126: llvm::LiveInterval& llvm::LiveIntervals::createEmptyInterval(unsigned int): Assertion '!hasInterval(Reg) && "Interval already exists!"' failed)`) and I'm not sure whether it used to. It still fails when I revert my changes, assuming the rebuild afterwards worked properly.  
This is on the `avr-rust-llvm-release-4-0-1` branch, with one change I made to fix the build; I doubt that causes this, though.

If anyone spots this, does that test work for you?

exscape commented:
I updated the commit with the requested fixes.  
I couldn't replicate the test failure (`high-pressure-on-ptrregs.ll`) on a clean build, so I think that part's OK.

I couldn't get the `CHECK-LABEL` statements to work properly however, as adding them in any fashion gave me errors about other checks not appearing, either because `FileCheck` started scanning too late (after the instruction in question), or (most of the time) stopped scanning too early. I tried every variation I could think of, but never got the tests to pass.  
I added regular `CHECK` statements for the labels instead; I hope that is acceptable.

dylanmckay commented:
> I added regular CHECK statements for the labels instead; I hope that is acceptable.

That's fine, to be honest I don't personally know the difference between `CHECK` and `CHECK-LABEL`, I've just seen the latter used for labels most of the time and have used that.

Using `CHECK` is fine 💯

dylanmckay commented:
Thanks for the pull request by the way!

I will merge both this and #4 and commit them to LLVM now

dylanmckay commented:
Ran tests locally, all pass

dylanmckay commented:
Upstreamed in [r314891](http://llvm.org/viewvc/llvm-project?view=revision&revision=314891)

dylanmckay commented:
Have created a merge request to get this into LLVM 5.0.1 [here](https://llvm.org/PR34829).
